File: ios/Tiny/Sources/TaskTree.swift
/**
 * TaskTree — native spawn_agents fan-out tree (web: TaskTree.tsx).
 *
 * Tool input arrives at beforeToolCallEvent (tasks known, all "running"),
 * the batch result lands with afterToolCallEvent and flips nodes to ✓/✗.
 * Rows expand to show each sub-agent's result text.
 */
import SwiftUI

struct SpawnNode: Identifiable, Equatable, Codable {
    var id: Int          // 1-based task number
    var prompt: String
    var ok: Bool?        // nil = running
    var result: String?
}

/// What one sub-agent's row is actually saying. `ok: Bool?` alone cannot say
/// it: `nil` means four different things depending on how the BATCH ended, and
/// three of them are not failure.
enum SpawnState: Equatable, CaseIterable {
    case running       // the batch is live and this one hasn't reported yet
    case queued        // launched in the background — it reports elsewhere, later
    case succeeded
    case failed
    case didNotRun     // the batch ended and never reported this task

    /// What VoiceOver says. Lives on the state, not the row (house shape:
    /// `DevicePeek.spoken`), because `didNotRun` and `failed` SHARE a glyph —
    /// dimmed vs full red — so these words are the only thing separating "we
    /// never started this" from "this one broke".
    var spoken: String {
        switch self {
        case .running: return "running"
        case .queued: return "queued"
        case .succeeded: return "succeeded"
        case .failed: return "failed"
        case .didNotRun: return "didn't run"
        }
    }
}

struct SpawnTreeItem: Identifiable, Equatable, Codable {
    /// How the batch itself ended — the missing half of every `ok == nil`.
    enum Outcome: String, Equatable, Codable {
        case live        // tasks announced, result not in yet
        case background  // wait:false — the fan-out was LAUNCHED, not finished
        case settled     // a results array arrived
        case aborted     // the tool errored, or its payload was unreadable
    }

    let id: String       // toolUseId
    var nodes: [SpawnNode]
    var elapsedMs: Double?
    var outcome: Outcome = .live

    /// ⚠️ Old persisted history has no `outcome`, and the default must NOT be
    /// `.live` there: a restored tree can never resume — its stream is gone —
    /// so a node that never reported didn't run, rather than spinning forever
    /// in a chat the user scrolled back to.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        nodes = try c.decode([SpawnNode].self, forKey: .nodes)
        elapsedMs = try c.decodeIfPresent(Double.self, forKey: .elapsedMs)
        outcome = try c.decodeIfPresent(Outcome.self, forKey: .outcome) ?? .settled
    }

    init(id: String, nodes: [SpawnNode], elapsedMs: Double?, outcome: Outcome = .live) {
        self.id = id
        self.nodes = nodes
        self.elapsedMs = elapsedMs
        self.outcome = outcome
    }

    /// The one place `ok` and `outcome` are read together. Pure, so the truth
    /// table is a test and not a screenshot.
    static func state(ok: Bool?, outcome: Outcome) -> SpawnState {
        if let ok { return ok ? .succeeded : .failed }
        switch outcome {
        case .live: return .running
        case .background: return .queued
        // Both are terminal for this node: nothing more is coming on this
        // stream, and it never reported. Saying "failed" would be a claim about
        // work the app never saw run.
        case .settled, .aborted: return .didNotRun
        }
    }

    func state(of node: SpawnNode) -> SpawnState { Self.state(ok: node.ok, outcome: outcome) }

    /// Parse the spawn_agents tool-result JSON: {ok, elapsed_ms, results:[{task, ok, result?, error?}]}
    mutating func apply(resultsJson: String) {
        guard let data = resultsJson.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            // Was a bare `return`, which left every node `nil` with the tree
            // still `.live` — the spinner that never stopped.
            outcome = .aborted
            return
        }
        // wait:false returns {ok, pending:true, batch_id, tasks} and NO results:
        // the batch is off and running, reporting later through a push and
        // use_device action:'result'. The old code read the absent results as a
        // total wipeout and painted every row red — "3 agents failed" about 3
        // agents that were all still working.
        if (obj["pending"] as? Bool) == true {
            outcome = .background
            return
        }
        elapsedMs = (obj["elapsed_ms"] as? NSNumber)?.doubleValue
        let results = obj["results"] as? [[String: Any]] ?? []
        for r in results {
            guard let task = r["task"] as? Int, let i = nodes.firstIndex(where: { $0.id == task }) else { continue }
            let ok = (r["ok"] as? Bool) ?? false
            nodes[i].ok = ok
            nodes[i].result = (r["result"] as? String) ?? (r["error"] as? String)
        }
        // An unreported task stays `nil` on purpose now. `.settled` makes it
        // read as "didn't run", which is all the app knows: the batch finished
        // and this one is simply not in the report.
        outcome = .settled
    }
}

struct TaskTreeCard: View {
    let item: SpawnTreeItem
    @Environment(\.tinyAccent) private var accent
    @State private var openIdx: Int?

    /// ⚠️ `.live` is the whole condition. Without it, a background batch (whose
    /// nodes never report on this stream) and an aborted one both spun forever.
    private var running: Bool { item.outcome == .live && item.nodes.contains { $0.ok == nil } }
    private var okCount: Int { item.nodes.filter { $0.ok == true }.count }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            header
            ForEach(item.nodes) { node in
                SpawnNodeRow(state: item.state(of: node),
                             node: node,
                             isLast: node.id == item.nodes.count,
                             isOpen: openIdx == node.id) {
                    openIdx = openIdx == node.id ? nil : node.id
                }
            }
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color.black.opacity(0.4), in: RoundedRectangle(cornerRadius: 14))
        .overlay(RoundedRectangle(cornerRadius: 14).stroke(accent.opacity(running ? 0.5 : 0.3), lineWidth: 1))
    }

    /// The right-hand summary, or nothing. ⚠️ `okCount` is only an answer once a
    /// report exists: "0/3 ok" beside a background batch is a scoreboard for a
    /// game that hasn't been played.
    private var summary: String? {
        switch item.outcome {
        case .live: return nil
        case .background: return "running in the background"
        case .aborted: return "ended without reporting"
        case .settled:
            guard let ms = item.elapsedMs else { return "\(okCount)/\(item.nodes.count) ok" }
            return "\(okCount)/\(item.nodes.count) ok · \(String(format: "%.1f", ms / 1000))s"
        }
    }

    private var header: some View {
        HStack(spacing: 6) {
            if running {
                ProgressView().scaleEffect(0.6)
            } else {
                // SF Symbol, not the 🤖 that was here: the house rule (and the
                // web's IconCpu) — and VoiceOver reads an emoji aloud as
                // "robot face" in the middle of a status line.
                Image(systemName: "cpu")
                    .font(.caption)
                    .foregroundStyle(accent)
                    .accessibilityHidden(true)
            }
            Text("spawn_agents · \(item.nodes.count) parallel")
                .font(.caption.weight(.semibold).monospaced())
                .foregroundStyle(accent)
            Spacer()
            if let summary {
                Text(summary)
                    .font(.caption2.monospaced())
                    .foregroundStyle(.secondary)
            }
        }
    }
}

private struct SpawnNodeRow: View {
    @Environment(\.tinyAccent) private var accent
    let state: SpawnState
    let node: SpawnNode
    let isLast: Bool
    let isOpen: Bool
    let toggle: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack(alignment: .top, spacing: 6) {
                // Decorative rail. Unhidden, VoiceOver announces the box-drawing
                // character before every prompt.
                Text(isLast ? "└" : "├")
                    .font(.caption.monospaced())
                    .foregroundStyle(accent.opacity(0.4))
                    .accessibilityHidden(true)
                Button(action: toggle) {
                    HStack(alignment: .top, spacing: 6) {
                        statusIcon
                        Text("#\(node.id) \(node.prompt)")
                            .font(.caption)
                            .foregroundStyle(state == .succeeded || state == .failed
                                             ? Color.primary : Color.secondary)
                            .lineLimit(isOpen ? nil : 1)
                            .multilineTextAlignment(.leading)
                    }
                }
                .disabled(node.result == nil)
                // The glyph carries the state, so it has to be spoken. Without
                // it a VoiceOver user hears every prompt and not one outcome —
                // the one thing that matters in a tree of parallel agents.
                .accessibilityLabel("#\(node.id) \(node.prompt), \(state.spoken)")
            }
            if isOpen, let r = node.result {
                Text(r)
                    .font(.caption)
                    .foregroundStyle(node.ok == true ? Color.primary.opacity(0.85) : Color.red)
                    .padding(8)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Color.black.opacity(0.3), in: RoundedRectangle(cornerRadius: 8))
                    .padding(.leading, 18)
            }
        }
    }

    @ViewBuilder
    private var statusIcon: some View {
        switch state {
        case .running:
            ProgressView().scaleEffect(0.5).frame(width: 12, height: 12)
        case .queued:
            // A dot, not a spinner: nothing is happening on this screen, and a
            // spinner promises an update that will arrive by push instead.
            Text("·").font(.caption.bold()).foregroundStyle(.secondary)
        case .succeeded:
            Text("✓").font(.caption.bold()).foregroundStyle(accent)
        case .failed:
            Text("✗").font(.caption.bold()).foregroundStyle(Color.red)
        case .didNotRun:
            // Dimmed, because the app is not claiming this one ran and broke.
            Text("✗").font(.caption.bold()).foregroundStyle(Color.red.opacity(0.55))
        }
    }
}

File: android/app/src/main/java/technology/tiny/app/ui/TaskTree.kt
package technology.tiny.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.AccountTree
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import org.json.JSONObject
import technology.tiny.app.ui.theme.TinyGray

/**
 * Native spawn_agents fan-out tree (iOS TaskTree.swift / web TaskTree.tsx).
 *
 * Tool input arrives at beforeToolCallEvent (tasks known, all "running"), the
 * batch result lands with afterToolCallEvent and flips nodes to ✓/✗. Rows expand
 * to show each sub-agent's result text.
 */
data class SpawnNode(
    val id: Int,          // 1-based task number
    val prompt: String,
    val ok: Boolean? = null, // null = running
    val result: String? = null,
)

data class SpawnTree(
    val id: String,       // toolUseId
    val nodes: List<SpawnNode>,
    val elapsedMs: Double? = null,
) {
    /** Parse the spawn_agents tool-result JSON: {ok, elapsed_ms, results:[{task, ok, result?, error?}]}. */
    fun applyResults(resultsJson: String): SpawnTree {
        val obj = runCatching { JSONObject(resultsJson) }.getOrNull() ?: return this
        val elapsed = if (obj.has("elapsed_ms")) obj.optDouble("elapsed_ms") else null
        val results = obj.optJSONArray("results")
        val byTask = HashMap<Int, Pair<Boolean, String?>>()
        for (i in 0 until (results?.length() ?: 0)) {
            val r = results!!.optJSONObject(i) ?: continue
            val task = r.optInt("task", -1)
            if (task < 0) continue
            val ok = r.optBoolean("ok", false)
            val text = r.optString("result").takeIf { it.isNotEmpty() }
                ?: r.optString("error").takeIf { it.isNotEmpty() }
            byTask[task] = ok to text
        }
        val updated = nodes.map { n ->
            byTask[n.id]?.let { (ok, text) -> n.copy(ok = ok, result = text) }
            // Anything unreported is a failure (batch timeout isolation).
                ?: n.copy(ok = n.ok ?: false)
        }
        return copy(nodes = updated, elapsedMs = elapsed)
    }
}

@androidx.compose.runtime.Composable
fun TaskTreeCard(tree: SpawnTree) {
    val running = tree.nodes.any { it.ok == null }
    val okCount = tree.nodes.count { it.ok == true }
    var openId by remember(tree.id) { mutableStateOf<Int?>(null) }

    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(Color.Black.copy(alpha = 0.4f))
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        // header
        Row(verticalAlignment = Alignment.CenterVertically) {
            if (running) {
                CircularProgressIndicator(Modifier.size(14.dp), strokeWidth = 2.dp, color = MaterialTheme.colorScheme.primary)
            } else {
                Icon(Icons.Outlined.AccountTree, contentDescription = null, tint = MaterialTheme.colorScheme.primary, modifier = Modifier.size(16.dp))
            }
            Spacer(Modifier.width(6.dp))
            Text(
                "spawn_agents · ${tree.nodes.size} parallel",
                fontFamily = FontFamily.Monospace,
                fontWeight = FontWeight.SemiBold,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.primary,
            )
            Spacer(Modifier.weight(1f))
            tree.elapsedMs?.let { ms ->
                val secs = ((ms / 100).toInt() / 10.0) // one decimal, no Locale/format dep
                Text(
                    "$okCount/${tree.nodes.size} ok · ${secs}s",
                    fontFamily = FontFamily.Monospace,
                    style = MaterialTheme.typography.labelSmall,
                    color = TinyGray,
                )
            }
        }
        tree.nodes.forEachIndexed { i, node ->
            SpawnNodeRow(
                node = node,
                isLast = i == tree.nodes.lastIndex,
                isOpen = openId == node.id,
                onToggle = { openId = if (openId == node.id) null else node.id },
            )
        }
    }
}

@androidx.compose.runtime.Composable
private fun SpawnNodeRow(node: SpawnNode, isLast: Boolean, isOpen: Boolean, onToggle: () -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        Row(verticalAlignment = Alignment.Top) {
            Text(
                if (isLast) "└" else "├",
                fontFamily = FontFamily.Monospace,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.primary.copy(alpha = 0.4f),
            )
            Spacer(Modifier.width(6.dp))
            Row(
                Modifier.let { if (node.result != null) it.clickable { onToggle() } else it },
                verticalAlignment = Alignment.Top,
            ) {
                StatusIcon(node.ok)
                Spacer(Modifier.width(6.dp))
                Text(
                    "#${node.id} ${node.prompt}",
                    style = MaterialTheme.typography.labelMedium,
                    color = if (node.ok == null) TinyGray else MaterialTheme.colorScheme.onSurface,
                    maxLines = if (isOpen) Int.MAX_VALUE else 1,
                )
            }
        }
        if (isOpen && node.result != null) {
            Text(
                node.result,
                style = MaterialTheme.typography.labelMedium,
                color = if (node.ok == true) MaterialTheme.colorScheme.onSurface.copy(alpha = 0.85f)
                else MaterialTheme.colorScheme.error,
                modifier = Modifier
                    .padding(start = 18.dp)
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(8.dp))
                    .background(Color.Black.copy(alpha = 0.3f))
                    .padding(8.dp),
            )
        }
    }
}

@androidx.compose.runtime.Composable
private fun StatusIcon(ok: Boolean?) {
    when (ok) {
        null -> CircularProgressIndicator(Modifier.size(12.dp), strokeWidth = 1.5.dp, color = TinyGray)
        true -> Text("✓", fontWeight = FontWeight.Bold, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.primary)
        false -> Text("✗", fontWeight = FontWeight.Bold, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.error)
    }
}
